examples/ct/handlers.go

```diff
@@ -100,454 +122,362 @@ type LogContext struct {
+	var method string
+	if isPrecert {
+		method = "AddPreChain"
+		signerFn = signV1SCTForPrecertificate
+	} else {
+		method = "AddChain"
+		signerFn = signV1SCTForCertificate
 	}
 
-	addChainRequest, err := parseBodyAsJSONChain(c, w, r)
-
+	// Check the contents of the request and convert to slice of certificates.
+	addChainReq, err := parseBodyAsJSONChain(c, r)
 	if err != nil {
-		return http.StatusBadRequest, err
+		return http.StatusBadRequest, fmt.Errorf("failed to parse add-chain body: %v", err)
 	}
-
-	// We already checked that the chain is not empty so can move on to verification
-	validPath, err := verifyAddChain(c, addChainRequest, w, isPrecert)
-
+	chain, err := verifyAddChain(c, addChainReq, w, isPrecert)
 	if err != nil {
-		// Chain rejected by verify.
-		return http.StatusBadRequest, err
+		return http.StatusBadRequest, fmt.Errorf("failed to verify add-chain contents: %v", err)
 	}
 
 	// Build up the SCT and MerkleTreeLeaf. The SCT will be returned to the client and
 	// the leaf will become part of the data sent to the backend.
-	var merkleTreeLeaf ct.MerkleTreeLeaf
-	var sct ct.SignedCertificateTimestamp
-
-	if isPrecert {
-		merkleTreeLeaf, sct, err = signV1SCTForPrecertificate(c.logKeyManager, validPath[0], c.timeSource.Now())
-	} else {
-		merkleTreeLeaf, sct, err = signV1SCTForCertificate(c.logKeyManager, validPath[0], c.timeSource.Now())
+	var issuer *x509.Certificate
+	if len(chain) > 1 {
+		issuer = chain[1]
 	}
-
+	merkleLeaf, sct, err := signerFn(c.logKeyManager, chain[0], issuer, c.timeSource.Now())
 	if err != nil {
-		return http.StatusInternalServerError, fmt.Errorf("failed to create / serialize SCT or Merkle leaf: %v %v", sct, err)
+		return http.StatusInternalServerError, fmt.Errorf("failed to build SCT and Merkle leaf: %v %v", sct, err)
 	}
 
-	// Inputs validated, pass the request on to the back end after hashing and serializing
-	// the data for the request
-	leaf, err := buildLogLeafForAddChain(c, merkleTreeLeaf, validPath)
-
+	// Send the Merkle tree leaf on to the Log server.
+	leaf, err := buildLogLeafForAddChain(c, merkleLeaf, chain)
 	if err != nil {
-		// Failure reason already logged
-		return http.StatusInternalServerError, err
+		return http.StatusInternalServerError, fmt.Errorf("failed to build LogLeaf: %v", err)
 	}
+	req := trillian.QueueLeavesRequest{LogId: c.logID, Leaves: []*trillian.LogLeaf{&leaf}}
 
-	request := trillian.QueueLeavesRequest{LogId: c.logID, Leaves: []*trillian.LogLeaf{&leaf}}
-
-	ctx, cancelFunc := context.WithDeadline(context.Background(), getRPCDeadlineTime(c))
-	defer cancelFunc()
-	response, err := c.rpcClient.QueueLeaves(ctx, &request)
-
-	if err != nil || !rpcStatusOK(response.GetStatus()) {
-		// TODO(Martin2112): Possibly cases where the request we sent to the backend is invalid
-		// which isn't really an internal server error.
-		// Request failed on backend
-		return http.StatusInternalServerError, err
+	if glog.V(2) {
```

Comment:
I was trying to avoid the string formatting [overhead](https://godoc.org/github.com/golang/glog#V); however, it's probably premature to optimize -- I'll shift to the shorter version and we can change it later if profiling shows any adverse affect.